arch/aarch64/mcount-dynamic.c

```diff
@@ -76,47 +82,168 @@ int mcount_setup_trampoline(struct mcount_dynamic_info *mdi)
 	return 0;
 }
 
+static void read_patchable_loc(struct mcount_dynamic_info *mdi,
+			    struct uftrace_elf_data *elf,
+			    struct uftrace_elf_iter *iter,
+			    unsigned long offset)
+{
+	typeof(iter->shdr) *shdr = &iter->shdr;
+
+	mdi->nr_patch_target = shdr->sh_size / sizeof(long);
+	mdi->patch_target = xmalloc(shdr->sh_size);
+
+	elf_get_secdata(elf, iter);
+	elf_read_secdata(elf, iter, 0, mdi->patch_target, shdr->sh_size);
+}
+
+void mcount_arch_find_module(struct mcount_dynamic_info *mdi,
+			     struct symtab *symtab)
+{
+	struct uftrace_elf_data elf;
+	struct uftrace_elf_iter iter;
+	unsigned int fpatchable_nop_patt[] = { 0xd503201f, 0xd503201f };
+	unsigned i = 0;
+
+	mdi->type = DYNAMIC_NONE;
+
+	if (elf_init(mdi->map->libname, &elf) < 0)
+		goto out;
+
+	elf_for_each_shdr(&elf, &iter) {
+		char *shstr = elf_get_name(&elf, &iter, iter.shdr.sh_name);
+
+		if (!strcmp(shstr, PATCHABLE_SECT)) {
+			mdi->type = DYNAMIC_PATCHABLE;
+			read_patchable_loc(mdi, &elf, &iter, mdi->base_addr);
+			goto out;
+		}
+	}
+
+	/* check first few functions have fentry signature */
+	for (i = 0; i < symtab->nr_sym; i++) {
+		struct sym *sym = &symtab->sym[i];
+		void *code_addr = (void *)sym->addr + mdi->map->start;
+
+		if (sym->type != ST_LOCAL_FUNC && sym->type != ST_GLOBAL_FUNC)
+			continue;
+
+		/* dont' check special functions */
+		if (sym->name[0] == '_')
+			continue;
+
+		/* only support calls to __fentry__ at the beginning */
+		if (!memcmp(code_addr, fpatchable_nop_patt, CODE_SIZE)) {
+			mdi->type = DYNAMIC_PATCHABLE;
```

Comment:
I got a bit confused. The current logic loops over the patchable locations from the section so it won't work without having the section even if we set the type to `DYNAMIC_PATCHABLE` here. I will check more if there is really a case where the section is removed.

So I just removed this logic in both x86_64 and aarch64.

Reply:
I just checked again, but the `__patchable_function_entries` section can be removed by `-Wl,--gc-sections`.
```
       --gc-sections
       --no-gc-sections
           Enable garbage collection of unused input sections.  It is
           ignored on targets that do not support this option.  The default
           behaviour (of not performing this garbage collection) can be
           restored by specifying --no-gc-sections on the command line.
           Note that garbage collection for COFF and PE format targets is
           supported, but the implementation is currently considered to be
           experimental.

           --gc-sections decides which input sections are used by examining
           symbols and relocations.  The section containing the entry symbol
           and all sections containing symbols undefined on the command-line
           will be kept, as will sections containing symbols referenced by
           dynamic objects.  Note that when building shared libraries, the
           linker must assume that any visible symbol is referenced.  Once
           this initial set of sections has been determined, the linker
           recursively marks as used any section referenced by their
           relocations.  See --entry and --undefined.

           This option can be set when doing a partial link (enabled with
           option -r).  In this case the root of symbols kept must be
           explicitly specified either by an --entry or --undefined option
           or by a "ENTRY" command in the linker script.
```
Maybe we better handle this case without saving original instruction, but I will leave it as a future work.

Reply:
I'm not sure but it might be difficult to keep only the `__patchable_function_entries` when `--gc-sections` is enabled. We might need to have a way to apply patching only to the `NOP`s rather than force applying all the full dynamic tracing patching. In this case, we don't have to worry about not having the section.

Reply:
This case is handled in https://github.com/namhyung/uftrace/pull/1367/commits/92ef7a86c5062a8432579ce570ad14e5ebb5ff4e and this PR is updated again.